Re: Rouge Score Implementation

gucifer opened the pull request:
Fixed Formatting Issue on my end. 
Added a simple Implementation for the ROUGE Metric.

Description:
Simple Implementation of the ROUGE metric (both Rouge-N and Rouge-L).
Check list:
- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

vfdev-5 reviewed (commented):
Thanks for updating @gucifer !
I left few initial comments without checking the implementation for now...

gucifer commented:
Any idea on how to get rid of this error:

```
error: Your local changes to the following files would be overwritten by checkout:
	ignite/metrics/rouge.py
Please commit your changes or stash them before you switch branches.
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,151 @@
+class Rouge(Metric):
+    r"""Calculates the Rouge Score for two Sequence of Tokens.
+
+
```
Cite the reference paper :

Lin, Chin-Yew. 2004. ROUGE: a Package for Automatic Evaluation of Summaries. In Proceedings of the Workshop on Text Summarization Branches Out (WAS 2004), Barcelona, Spain, July 25 - 26, 2004.

http://www.aclweb.org/anthology/W/W04/W04-1013.pdf

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,151 @@
+    r"""Calculates the Rouge Score for two Sequence of Tokens.
+
+
+    .. math::
```
I suppose Rouge-L is considered here. Formulas don't match the reference paper. Is it normal ?

gucifer replied:
The one in the paper is only ROUGE- recall. Historically, ROUGE-recall and BLEU-precision were used, but later on ROUGE-precision and ROUGE-f1-score were introduced. Currently ROUGE-f1 average is the most widely used metric.
Also, I will add example and test cases for the metric in the next commit.

gucifer replied:
Is a simple example like this one fine?
```
from ignite.metrics import Rouge
import torch
m = Rouge(alpha=1,n='l')
peer = "the cat was found under the bed"
model = "the cat was under the bed"
m.update([peer.split(),[model.split()]])
peer = "the tiny little cat was found under the big funny bed"
model = "the cat was under the bed"
m.update([peer.split(),[model.split()]])
print(m._num_examples)
print(m.compute())
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,151 @@
+            form expected by the metric. This can be useful if, for example, you have a multi-output model and
+            you want to compute the metric with respect to one of the outputs.
+
+
```
add 
```
.. versionadded:: 0.5.0
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,151 @@
+
+    """
+
+    def __init__(self, alpha: float, n: int, device: Union[str, torch.device] = torch.device("cpu")) -> None:
```
add argument to `__init__`
```
output_transform: Callable = lambda x: x
```

gucifer replied:
I will. Also, What is the use of that?

sdesrozis replied:
A way for users to transform output from `Engine` that could be not suitable for the metric. 

See https://pytorch.org/ignite/master/metrics.html#ignite-metrics

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,151 @@
+    def _check_parameters(self, alpha: float, n: int) -> None:
+        if alpha < 0 or alpha > 1:
+            raise ValueError("Alpha has to be between 0 and 1 to calculate Rouge f1 score.")
+        if isinstance(n, str) and n != "l" and n != "L":
```
if `n` could be a string, the type in signature should be `Union[str, int]` and the docstring should be updated to explain the usage.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,151 @@
+        elif isinstance(n, int) and n < 1:
+            raise ValueError("Ignite needs atleast unigram to calculate Rouge")
+
+    def _ngramify(self, text: List[str], n: int) -> Dict:
```
```suggestion
    def _ngramify(self, tokens: List[str], n: int) -> Dict:
```
Better name for this argument ??

gucifer replied:
Both are pretty good. I will change those.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,151 @@
+
+
+    .. math::
+        F_\1 = \left( 1 - \alpha \right) * \frac{ \text{precision} * \text{recall} }
```
```suggestion
        F_\alpha = \left( 1 - \alpha \right) * \frac{ \text{precision} * \text{recall} }
```

gucifer replied:
f1 = (precision*recall)/((1-alpha)*precision + alpha*recall)
This is the actual formula. My mathematical notation could be wrong.

sdesrozis replied:
See for instance https://en.wikipedia.org/wiki/F-score

`f1` can't depend on a parameter `alpha`. In original paper, the author uses `f_\beta` score.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,151 @@
+        else:
+            return 0.0
+
+    def _f1_score(self, matches: int, recall_total: int, precision_total: int, alpha: float = 1) -> float:
```
if `alpha = 0.5` we get `0.5 f_1`
if `alpha = 1` we get `precision`
if `alpha = 0` we get `recall`
This is not the `f_\alpha` in the reference paper, right ?

I think the reference implementation is https://github.com/bdusell/rougescore and this is a variation of the original paper. Do we want this or original algorithm ? I prefer provide the original algorithm.

gucifer replied:
Implementation is inline with  https://github.com/google-research/google-research/blob/dbdf11998a865ba719b651310675f530c3c5763a/rouge/rouge_scorer.py#L125

gucifer replied:
Only difference is the preprocessing between the  https://github.com/bdusell/rougescore  and the google research implementation. And we are leaving the preprocessing of text to user that's why we are taking sequence of tokens as input and not the original text. also at alpha = 0.5 we get original f1-score, but sometimes we might want more weight towards precision or recall that is why alpha is provided.

sdesrozis replied:
https://github.com/google-research/google-research/blob/dbdf11998a865ba719b651310675f530c3c5763a/rouge/scoring.py#L162

Your implementation of `f1` is different.

gucifer replied:
what I am saying is alpha = 0.5 is that formula. but you can change alpha depending on which is more important precision or recall. If you want I can remove alpha, and just add simple 2* (formula) instead. I thought giving slight flexibility to the user would be nice, since it does not require much code and would be nicer for a user. Also, I have deliberately made alpha a keyword argument so that if a user does not add alpha we can by default use the 2*(formula) instead.

sdesrozis replied:
IMHO we should provide a clean implementation of the reference paper first and use right term for scores that we use.

sdesrozis commented on ignite/metrics/rouge.py.orig:
```diff
@@ -0,0 +1,152 @@
+from ignite.metrics import Metric
```
Please do not commit useless files

gucifer replied:
Sorry, I had formatting issue on my side and generated this file while trying to fix that. was not aware of it.

sdesrozis commented:
@gucifer I left comments. IMO we should provide an implementation of the original paper. What do you think ?

gucifer commented:
> @gucifer I left comments. IMO we should provide an implementation of the original paper. What do you think ?

Can we discuss this over discord? Can you provide me your discord?

sdesrozis commented:
Sorry I have any discord yet. I will create an account as soon as possible to discuss

gucifer commented:
> Sorry I have any discord yet. I will create an account as soon as possible to discuss
> 
> Edit :
> Discord id
> Sylvain Desroziers#6973

Sent you a Friend Request. My id is kana#7979

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+
+
+    .. math::
+        F_\1 = 2 * \frac{\text{precision} * \text{recall}} {\text{precision} + \text{recall}}
```
If f1 is used, remove useless alpha parameter.

But the paper relies on fbeta, so I think alpha should be renamed beta and implement fbeta formula rather that f1.

As discussed, we could think about variant later.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+
+    Args:
+        alpha: alpha value for calculating f1 score
+        n: Rouge score to be calculated using n-grams of tokens
```
Int values relies on n-gram. String values should be explained.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+    @reinit__is_reduced
+    def update(self, output: List[List]) -> None:
+        y_pred, y = output[0], output[1]
+        if self.n == 0:
```
Suggestion
In `__init__`
```python 
self.rouge_fn = self.rouge_l if n == 0 else self.rouge_n
```
Therefore
```python

self._rougetotal = torch.add(self._rougetotal, self.rouge_fn(y_pred, y))
```
without test.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+        else:
+            return 0.0
+
+    def _f1_score(self, matches: int, recall_total: int, precision_total: int) -> float:
```
As mentioned above, I suggest fbeta.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+
+    Example:
+        from ignite.metrics import Rouge
+        import torch
```
```suggestion
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+        y_pred = "the tiny little cat was found under the big funny bed"
+        y = "the cat was under the bed"
+        m.update([y_pred.split(),[y.split()]])
+        print(m._rougetotal)
```
```suggestion
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+        y = "the cat was under the bed"
+        m.update([y_pred.split(),[y.split()]])
+        print(m._rougetotal)
+        print(m._num_examples)
```
```suggestion
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+        y = "the cat was under the bed"
+        m.update([y_pred.split(),[y.split()]]) #Using space to separate sentences into tokens
+        y_pred = "the tiny little cat was found under the big funny bed"
+        y = "the cat was under the bed"
```
```suggestion
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+    ) -> None:
+        self._rougetotal: torch.Tensor = torch.tensor(0, device=device)
+        self._num_examples: int = 0
+        self._check_parameters(n)
```
```suggestion
        self.n = self._check_parameters(n)
```
The function should return the valid int value. No rule outside this function.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,157 @@
+        self.n: int = (0 if isinstance(n, str) else n)
+        super(Rouge, self).__init__(output_transform=output_transform, device=device)
+
+    def _check_parameters(self, n: Union[int, str]) -> None:
```
```suggestion
    def _check_parameters(self, n: Union[int, str]) -> int:
```
and
```python
return 0 if isinstance(n, str) else n
```

sdesrozis commented:
@gucifer Thank you for the update. I left comments.

Tests would be added.

vfdev-5 commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,158 @@
+    @reinit__is_reduced
+    def update(self, output: List[List]) -> None:
+        y_pred, y = output[0], output[1]
+        self._rougetotal = torch.add(self._rougetotal, self.rouge_fn(y_pred, y))
```
```suggestion
        self._rougetotal += self.rouge_fn(y_pred, y)
```

sdesrozis replied:
You did not fix this comment.

gucifer replied:
It will not work because the function output is a float and self._rougetotal is a tensor. See https://github.com/pytorch/ignite/pull/1650/commits/2333f5c27d1b32891ea5cce38cae277d895221fb

sdesrozis replied:
Ok thanks to point that.

vfdev-5 commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,158 @@
+        self._rougetotal = torch.add(self._rougetotal, self.rouge_fn(y_pred, y))
+        self._num_examples += 1
+
+    @sync_all_reduce("_num_examples", "_num_correct")
```
In sync_all_reduce we specify class attributes to call all_reduce on. You defined `_num_examples` but there is no `_num_correct`, right ? You can remove it and put `_rougetotal` instead.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,158 @@
+                diag = up
+        return left
+
+    def rouge_n(self, y_pred: List[str], y: List[List[str]]) -> float:
```
Why list of list for y ?

vfdev-5 commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,158 @@
+    def compute(self) -> torch.Tensor:
+        if self._num_examples == 0:
+            raise NotComputableError("Rouge must have at least one example before it can be computed.")
+        return torch.div(self._rougetotal, self._num_examples)
```
```suggestion
        return self._rougetotal / self._num_examples
```

sdesrozis commented:
@gucifer do you plan to continue on this issue ?

gucifer reviewed (commented):
> @gucifer do you plan to continue on this issue ?

Yes, I have my mid-term presentation tomorrow. I have taken a week off because of it.

sdesrozis commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,195 @@
+        assert dev == metric_device, f"{dev} vs {metric_device}"
+
+
+# def test_accumulator_detached():
```
@gucifer please remove useless code.

gucifer replied:
This is another question I wanted to ask. What does this function exactly do in other metrics and is it useful for rouge score?

sdesrozis replied:
This test checks that your tensors inside the metric do not need gradients. It would be relevant to have such validation.

gucifer replied:
Let me add that function to the tests then.

sdesrozis reviewed (approved):
@gucifer LGTM !

sdesrozis commented:
@vfdev-5 last chance to make some comments !!

sdesrozis commented:
@gucifer I was thinking about improve the test using seq2seq rouge comparison. What do you think ?

@vfdev-5 could we improve the tests in a new PR ?

gucifer commented:
> @gucifer I was thinking about improve the test using seq2seq rouge comparison. What do you think ?
How would seq2seq model affect the test?

sdesrozis commented:
I mean compare in test with this implementation (for instance) as we do wrt sklearn in others metrics 
https://github.com/google/seq2seq/blob/master/seq2seq/metrics/rouge.py

If you can do it in this PR, that would be great. Otherwise let’s see in another PR

gucifer commented:
> I mean compare in test with this implementation (for instance) as we do wrt sklearn in others metrics
> https://github.com/google/seq2seq/blob/master/seq2seq/metrics/rouge.py
> 
> If you can do it in this PR, that would be great. Otherwise let’s see in another PR

The test_compute method has values from that only. I did not import and use the rouge module because it would require separate installation.

sdesrozis commented:
> The test_compute method has values from that only. I did not import and use the rouge module because it would require separate installation.
> 
> For single sentences without special characters and punctuation the score of both would be 100% identical. For the rest of sentences, it may vary depending on the type of pre-processing and tokenization used. For the same pre-processing it will give the same result though.

It could be a requirement added to dev only https://github.com/pytorch/ignite/blob/master/requirements-dev.txt

Otherwise, what do you think about wrapping (sentence, scores) in a specific structure (dict, json, etc.) rather than in hand coded way ? 

My favorite option is to evaluate the metric as we do with sklearn.

gucifer commented:
> > The test_compute method has values from that only. I did not import and use the rouge module because it would require separate installation.
> > For single sentences without special characters and punctuation the score of both would be 100% identical. For the rest of sentences, it may vary depending on the type of pre-processing and tokenization used. For the same pre-processing it will give the same result though.
> 
> It could be a requirement added to dev only https://github.com/pytorch/ignite/blob/master/requirements-dev.txt
> 
> Otherwise, what do you think about wrapping (sentence, scores) in a specific structure (dict, json, etc.) rather than in hand coded way ? By the way, I agree to restrict the comparison to similar features.
> 
> My favorite option is to evaluate the metric as we do with sklearn.

yes then I will replace numeric scores with the rouge function from module. And wrapping scores in a dict, that is just a small edit so it is very doable as well. I can add both those things tommorow.

gucifer commented:
> Otherwise, what do you think about wrapping (sentence, scores) in a specific structure (dict, json, etc.) rather than in hand coded way ? By the way, I agree to restrict the comparison to similar features.
>

How do you want me to wrap it? A list of dictionaries where each dictionary has one training sample and it's score?

vfdev-5 commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,195 @@
+    y_preds = [["Hi there!", "Hi , How are you?"], ["Hello there", "Hello , How are you?"]]
+
+    def update(engine, i):
+        return (y_true[i].split(), [s.split() for s in y_preds[i]])
```
Why do we change now the order of `y` and `y_true` ?  Normally, we should have: `(y_pred, y)`

vfdev-5 reviewed (commented):
Thanks for the update @gucifer , but I still have few comments here before merging it.

sdesrozis commented:
@gucifer 

I did the following test
```
y_pred = "the cat was found under the bed"
y = "the cat was under the bed"

# your implementation
Rouge(n=2).rouge_n(y_pred.split(), [y.split()])
# => 0.6666666666666666

# from seq2seq
# https://github.com/google/seq2seq/blob/master/seq2seq/metrics/rouge.py
rouge_n([y_pred], [y])
# => 0.7272727223140496
```

Could you reproduce it ?

sdesrozis reviewed (changes requested):
Sorry further validation wrt others implementations is needed.

gucifer commented:
> Sorry further validation wrt others implementations is needed.

They are using this formula to calculate.   f1_score = 2.0 * ((precision * recall) / (precision + recall + 1e-8))
We are using f_beta. For f_beta to be equal to f1_score mentioned above, you must put in beta=1.0 as an argument as by default it is 0.

Try running Rouge(n=2,beta=1.0).rouge_n(y_pred.split(), [y.split()])

sdesrozis commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,195 @@
+    assert rouge.compute() == 0.7619047619047619
+
+
+def _test_distrib_integration(device):
```
I realized that this test does not check any numeric values.
This test should be refactor to be compatible with distributed validation.

The idea is to have a corpus replicated on each process. Each process work on a part of the corpus and compare the result of the metric with the local computation done.

See for instance https://github.com/pytorch/ignite/blob/455a3b3282eca7e15d45c22f7e4e384faef465fa/tests/ignite/metrics/test_precision.py#L784

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,168 @@
+            raise ValueError('Invalid String, Only Rouge-L supported.Please use "l" or "L"')
+        elif isinstance(n, int) and n < 1:
+            raise ValueError("Ignite needs atleast unigram to calculate Rouge")
+        return (beta, n)
```
```suggestion
        return beta, n
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,168 @@
+        super(Rouge, self).__init__(output_transform=output_transform, device=device)
+
+    def _check_parameters(self, beta: float, n: Union[int, str]) -> Tuple:
+        if not isinstance(beta, float):
```
```suggestion
        if not isinstance(beta, numbers.Number):
```
It would work with integral numbers.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,168 @@
+        self.rouge_fn = self.rouge_l if isinstance(self.n, str) else self.rouge_n
+        super(Rouge, self).__init__(output_transform=output_transform, device=device)
+
+    def _check_parameters(self, beta: float, n: Union[int, str]) -> Tuple:
```
It could be a static function.

gucifer replied:
The check_parameters function?

sdesrozis replied:
```suggestion
    def _check_parameters(beta: float, n: Union[int, str]) -> Tuple:
```
No `self` if static

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,168 @@
+
+def _ngramify(tokens: List[str], n: int) -> DefaultDict:
+    ngrams: DefaultDict = defaultdict(int)
+    for ngram in (tuple(tokens[i : i + n]) for i in range(len(tokens) - n + 1)):
```
```suggestion
    for ngram in (tuple(tokens[i: i + n]) for i in range(len(tokens) - n + 1)):
```
PEP

vfdev-5 replied:
Black can also "fix" spaces like that for indexing...

gucifer replied:
VScode pep8 formatter does not seem to fix it on its own. I will look into it.

vfdev-5 replied:
Well, let's keep it like that without any change. For instance, we do not configure black and just use it as is...

sdesrozis commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,266 @@
+        rouge.compute()
+
+
+def test_input():
```
```suggestion
def test_wrong_input():
```

sdesrozis commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,266 @@
+from ignite.metrics.rouge import _lcs
+
+
+def test_zero_div():
```
Merge in `test_wrong_inputs`

sdesrozis commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,266 @@
+        rouge = Rouge(n=-1)
+
+
+def testrougeempty():
```
```suggestion
def test_empty():
```
Please use PEP8 function naming rules
https://www.python.org/dev/peps/pep-0008/

sdesrozis commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,266 @@
+
+
+def testrougeagainstrouge155():
+    y_pred = """rFDJCRtion Ht-LM EKtDXkME,yz'RBr q'wer wrojNbN wL,b .a-'XdQggyFl jB-RPP'iyOIcUxi n cw-WeFyu vC MoBL Xdn g
```
Why not generate the texts using a repeating letter ?
```python
# generate 25 tokens of random size (1,10)
y_pred = " ".join(["a" * random.randint(1, 10) for _ in range(25)])
```

sdesrozis commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,266 @@
+    assert rougel.compute() == scores["rougeL"].fmeasure
+
+
+def test_compute():
```
This test validates the accumulation. Why not use `rouge-score` ?

sdesrozis replied:
This test should be refactored in a more compact way
```
def _test_compute(rouge, rouge_scorer):
    num_examples = 0
    acc_score = 0
    y_pred_list = ["the cat was found under the bed", "the tiny little cat was found under the big funny bed"]
    y_list = ["the cat was under the bed", "the cat was under the bed"]
    for i, (y_pred, y) in enumerate(zip(y_pred_list, y_list), start=1):
        score = scorer.score(y_pred, y)
        acc_score += score["rouge1"].fmeasure
            rouge.update([tokenize.tokenize(y_pred, None), tokenize.tokenize(y, None)])
        assert isinstance(rouge.compute(), torch.Tensor)
        assert rouge.compute() == acc_score / i

    _test_compute(Rouge(n=1, beta=1.0), rouge_scorer.RougeScorer(["rouge1", "rougeL"], use_stemmer=False))
```

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,178 @@
+    Args:
+        beta: beta value for calculating f-beta
+        variant: Which variant of Rouge Score to be evaluated
+            Valid Values - "rougeN", "rougen", "n", "N", "rougeL", "rougel", "l", "L"
```
This is not understandable... Too many ways to define the same algorithm.

For instance (think about the full version) 
```suggestion
            Valid Values - "corpus-N", "sentence-N", "corpus-L", "sentence-L", 
```
Some arguments should be mutual exclusive.

sdesrozis replied:
Please have a look of the notebook I send yesterday.

sdesrozis commented on ignite/metrics/rouge.py:
```diff
@@ -0,0 +1,178 @@
+        self,
+        beta: float = 0.0,
+        variant: str = "rougeN",
+        n: int = 1,
```
Useless if `N` is encoded in variant 
```
m = Rouge(beta=1,variant="rouge-4")  # here N == 4
```

sdesrozis commented on tests/ignite/metrics/test_rouge.py:
```diff
@@ -0,0 +1,287 @@
+        data = list(range(n_iters))
+        engine.run(data=data, max_epochs=1)
+
+        assert "rouge" in engine.state.metrics
```
May we use rouge-scorer to validate?
